File: 0001-0500/0129. Sum Root to Leaf Numbers.py
# Time:O(n), Space:O(h)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def dfs(node, cur_num):
            if not node:
                return 0

            cur_num = cur_num * 10 + node.val

            if not node.left and not node.right:
                return cur_num

            return dfs(node.left, cur_num) + dfs(node.right, cur_num)

        return dfs(root, 0)

File: 0001-0500/0129. Sum Root to Leaf Numbers.java
// Medium
// DFS
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/sum-root-to-leaf-numbers/

class Solution {
    public int sumNumbers(TreeNode root) {
        return dfs(root, 0);
    }
    
    private int dfs(TreeNode node, int curSum) {
        if (node == null)
            return 0;
        
        curSum = curSum * 10 + node.val;

        if (node.left == null && node.right == null) {
            return curSum;
        }

        int leftSum = dfs(node.left, curSum);
        int rightSum = dfs(node.right, curSum);
        return leftSum + rightSum;
    }
}
